Machine-readable output

m-aXimilian opened the pull request:
# ToDos
- [x] add schema for xml and json
- [x] add tests (i.e., "basic output works")
- [ ] add schema validation tests
- [x] add documentation

## Description

This PR introduces a new encoding switch to choose how the CZICheck's result should be emitted to the output.
The switch is exposed as a command line parameter `--encoding` (`-e` for short).
One can choose between `text` (the default and the status quo, i.e., human readable text), `json` which will emit a json encoded result, and `xml` which will emit xml-encoded output.
To achieve this, the dependencies [rapidjson](https://rapidjson.org/) and [pugiXML](https://pugixml.org/) are introduces to the project.

_Note_: libxo, as suggested in issue #16 has some shortcomings regarding cross-platform builds (Windows especially). This is why other libraries were used.

Fixes #16 , #23

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Describe the tests that you ran to verify your changes.  
Provide instructions to reproduce.

## Checklist:

- [x] I followed the Contributing Guidelines.
- [x] I did a self-review.
- [x] I commented my code, particularly in hard-to-understand areas.
- [x] I updated the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

ptahmose commented:
[version-number](https://github.com/ZEISS/czicheck/blob/44e0bc31f99e66bc6a77c5f2ef517c62a6728dec/CMakeLists.txt#L8C15-L8C20) and [version history](https://github.com/ZEISS/czicheck/blob/44e0bc31f99e66bc6a77c5f2ef517c62a6728dec/documentation/version-history.md) need a bump

ptahmose commented on CZICheck/CZICheck.cpp:
```diff
@@ -54,25 +54,25 @@ int main(int argc, char** argv)
-                case CResultGatherer::AggregatedResult::OK:
-                    log->WriteStdOut("\n\nResult: OK\n");
+                case IResultGatherer::AggregatedResult::OK:
+                    // log->WriteStdOut("\n\nResult: OK\n");
```
can be removed?

ptahmose commented on CZICheck/resultgathererfactory.h:
```diff
@@ -0,0 +1,29 @@
+#include "resultgathererxml.h"
+
+
+static std::unique_ptr<IResultGatherer> CreateResultGatherer(const CCmdLineOptions& options) {
```
Defining **and** declaring a static function in a header I'd see as problematic or questionable.

Declaring a function as static in a header file limits its linkage to the file where it’s included. This means that each translation unit that includes this header will get its own separate copy of the function. This can lead to increased code size if the header is included in multiple source files.

m-aXimilian commented on documentation/assets/json-output-schema.json:
This (as well as the xml-output-schema.xsd) documents the intention to use it for validating the json (and xml respectively) output against at some point (in test which is WIP currently).

ptahmose commented on CZICheck/resultgathererxml.h:
```diff
@@ -0,0 +1,54 @@
+    void FinalizeChecks() override;
+
+private:
+    static std::wstring ConvertToWideString(const std::string& str)
```
I'd suggest using [convertUtf8ToUCS2](https://github.com/ZEISS/czicheck/blob/44e0bc31f99e66bc6a77c5f2ef517c62a6728dec/CZICheck/utils.h#L13C1-L13C56)

m-aXimilian replied:
Oh nice, thanks! I was not aware of that...should have checked I guess :smile:

ptahmose commented on CZICheck/checks.h:
```diff
@@ -57,3 +58,30 @@ enum class CZIChecks
     /// The Applicance Metadata specified for TopographyDataItem(s) are valid
```
typo -> Appliance

ptahmose commented on CZICheck/resultgathererjson.cpp:
```diff
@@ -0,0 +1,116 @@
+    this->test_results_ = rapidjson::Value(rapidjson::kArrayType);
+}
+
+
```
extra line break

ptahmose commented on CZICheck/checks.h:
```diff
@@ -57,3 +58,30 @@ enum class CZIChecks
     ApplianceMetadataTopographyItemValid,
 };
+
+constexpr const char* CZIChecksToString(CZIChecks czi_check)
```
I'd recommend to not have the implementation in the header.

* Although it is declared constexpr, it may not often be used in a compile-time context (especially if CZIChecks is determined at runtime).
* If it cannot be evaluated at compile time, then every translation unit that includes this header would get its own copy of the function, increasing overall code size.
* It throws an exception, which is a runtime operation and might not align well with constexpr usage in practice, as constexpr functions ideally shouldn’t have side effects like exceptions.

m-aXimilian replied:
Makes sense! Ditched the `constexpr` an the header implementation. Thanks!

ptahmose commented:
By chance, I ran across [this](https://github.com/openzfs/zfs/pull/16217) which is about "something similar".
One idea I got from there - adding a "version info" to the output seems like a good idea to me.
Something like

```
"output_version": {
  "command": "CZICheck"
  "version": "1.2.3"
}
```

What do you think?

m-aXimilian commented:
I'd say, adding a version info to the output is a good idea. Especially when thinking about bugs that may haunt us in (the near) future, having less ambiguity in the output is definitely a valid point. I will do some fiddling integrating this. (For the xml output as well I would say).

m-aXimilian commented:
> By chance, I ran across [this](https://github.com/openzfs/zfs/pull/16217) which is about "something similar". One idea I got from there - adding a "version info" to the output seems like a good idea to me. Something like
> 
> ```
> "output_version": {
>   "command": "CZICheck",
>   "version": "1.2.3"
> }
> ```
> 
> What do you think?

Added that to the json and xml outputs. Please note: also updated the documentation and the json+xml schema accordingly.

ptahmose reviewed (commented):
LGTM!
Please consider my remarks and suggestions.

ptahmose reviewed (approved):
LGTM

m-aXimilian commented on CZICheck/resultgathererxml.cpp:
```diff
@@ -0,0 +1,126 @@
+
+    this->root_node_.append_child(L"AggregatedResult")
+        .text().set(ConvertToWideString(ss.str()).c_str());
+    this->xml_document_.save(std::cout, L" ");
```
Done in [this update](https://github.com/ZEISS/czicheck/pull/22/commits/ccb10412ee58129b337453b2c76cfd0713877ce8)

m-aXimilian commented on CZICheck/resultgathererxml.cpp:
```diff
@@ -0,0 +1,126 @@
+void CResultGathererXml::FinalizeChecks()
+{
+    ostringstream ss;
+    switch (this->GetAggregatedResult()) {
```
removed in [this update](https://github.com/ZEISS/czicheck/pull/22/commits/6a837d49815e9dbba4483b68808ca05d1bd1588c)

m-aXimilian commented on CZICheck/resultgathererxml.cpp:
```diff
@@ -0,0 +1,126 @@
+
+void CResultGathererXml::StartCheck(CZIChecks check)
+{
+    const auto checker_display_name = CCheckerFactory::GetCheckerDisplayName(check).c_str();
```
removed in [this update](https://github.com/ZEISS/czicheck/pull/22/commits/6a837d49815e9dbba4483b68808ca05d1bd1588c)

m-aXimilian commented on CZICheck/resultgathererxml.cpp:
```diff
@@ -0,0 +1,126 @@
+    for (auto current_test_node : this->test_node_.children())
+    {
+        auto name_attr = current_test_node.attribute(kTestNameId);
+        if (name_attr && utf8_conv.to_bytes(name_attr.value()) == this->current_checker_id_)
```
done in [this update](https://github.com/ZEISS/czicheck/pull/22/commits/adf27ae442bb2f63a816b554e9c1407fdd5a235d)

m-aXimilian commented on CZICheck/IResultGatherer.h:
```diff
@@ -0,0 +1,114 @@
+#include <string>
+#include "checks.h"
+
+class IResultGatherer
```
It started out as a "pure" interface but I degraded it to a (abstract) base class for practicality reasons. If it's ok for you, I'd leave the naming as is for now.

m-aXimilian commented on CZICheck/cmdlineoptions.cpp:
```diff
@@ -249,6 +286,35 @@ CCmdLineOptions::ParseResult CCmdLineOptions::Parse(int argc, char** argv)
+    if (encoding == nullptr)
+    {
+        return true;
+    }
```
fixed in [this update](https://github.com/ZEISS/czicheck/pull/22/commits/173558254749ca2568c80ec4fb59433b3a7e2515)

ptahmose commented on CZICheck/cmdlineoptions.cpp:
```diff
@@ -249,6 +287,38 @@ CCmdLineOptions::ParseResult CCmdLineOptions::Parse(int argc, char** argv)
+    {
+        error_message = "No output encoding was specified.";
+        return false;
+    }
```
I guess you aim for a different semantic than the other "Parse"-methods. I mean - you seem to consider a call with "encoding=nullptr" as an invalid call. That's fine with me, I'd just suggest - when considering "encoding" as a required parameter, then a reference would be preferable (and make this intention clear).

m-aXimilian replied:
This is true. And I initially thought about using a ref instead but decided against it to keep the "schema" aligned to the other parsers.
But since this makes the _if nullptr_ statement obsolete and less code is good code, I changed that in [this update](https://github.com/ZEISS/czicheck/pull/22/commits/4f39595a5af0730179a8407d596a2826379ae240)

m-aXimilian commented:
I have updated the test script to check (and create) expected results for json and XML output as well in [this update](https://github.com/ZEISS/czicheck/pull/22/commits/28c091d7178ed2bc8ca223a1ec7ff64e541e5aa2)
and the respective "good results" in [this update](https://github.com/ZEISS/czicheck/pull/22/commits/29bf9fc133b3356d595ff8bb998a21e6bb481c28)
@ptahmose what do you think?

m-aXimilian commented on CZICheck/resultgathererjson.cpp:
```diff
@@ -0,0 +1,116 @@
+    const auto& result = it->second;
+
+    auto allocator = this->json_document_.GetAllocator();
+    for (int res { 0 }; res < this->test_results_.Size(); ++res)
```
Will update this in the next PR
